feat: Configured subsection prerequisites persist during course olx/xml export/import

haftamuk opened the pull request:
# Configured subsection prerequisites persist during course olx/xml export/import

## Overview

This PR introduces support for subsection (sequential) prerequisites in the OLX course format. Course authors can define prerequisite relationships between subsections with configurable completion criteria including minimum score and completion percentage requirements. Currently this information is lost during course import/export. This PR ensures configured subsection prerequisites are persisted during course olx/xml export/import. It allows course authors to define prerequisite relationships, have them exported to OLX format, and finally, have them properly imported.


## Problem Statement
Currently, Open edX supports to define prerequisite relationships between subsections in Studio. This enables course authors from enforcing learning pathways where students must complete certain subsections before accessing others, which is crucial for structured learning experiences. But when a course is exported (XML/OLX) all configured relationships of subsections are lost and authors has to do it again manually in studio. This is time consuming and error prone!
 
Proposed Solution
- Extend the <sequential> element in OLX to support prerequisite definitions
- Maintain full backward compatibility with existing OLX courses

## Key Files

1. **Service Layer** (`xmodule/seq_block.py`)
   - Adds prerequisite information to sequential XML for export.

2. **Import Processing Layer** (`xmodule/modulestore/xml_importer.py`)
   - Processes prerequisite relationships during course import

## OLX Format

### Prerequisite information attached to sequential xml element

```xml
<sequential 
required_content="91d0290972c4488db10d7ca3694e13ca" 
min_score="100" 
min_completion="100" 
... >
    <vertical url_name="some_vertical"/>
   ...
</sequential>
```

## Subsection prerequisite configuration how-to
1. `Settings` -> `Advanced` -> `Enable subsection prerequisites` -> `true`
<img width="1079" height="114" alt="image" src="https://github.com/user-attachments/assets/44ac5150-67e8-4255-85df-142939c12ebd" />

**Note :**  Even after configuring a complete learning path for a course you can disable the gating feature by simply turning this in to false. Also, if you switch it back to true, all the previous configured learning path will appear with out further work. This means, switching this setting to true/false will not impact persisted learning path configuration, but will impact how the learner will interact with the course.

**Note :** When you export a course with this flag set to true, the exported XML will contain the information
```
<course cert_html_view_enabled="true" discussions_settings="{&quot;enable_graded_units&quot;: false, &quot;enable_in_context&quot;: true, &quot;provider_type&quot;: &quot;openedx&quot;, &quot;unit_level_visibility&quot;: true}" display_name="Saturday1"

enable_subsection_gating="true" 
end="null" language="en" start="2030-01-01T00:00:00Z">
  <chapter url_name="ca81aeb19a7f41a49f05a87c95fe299a"/>
  <chapter url_name="4d13584e169241bb890871bd44059d43"/>
  <wiki slug="Saturday1.Saturday1.Saturday1"/>
</course>

```
Later during import this setting will be applied to the destination course. This is also the case when it is set to false (the destination course will be `enable_subsection_gating` set to false)

2. Make a subsection to serve as a prerequisite
<img width="1126" height="630" alt="image" src="https://github.com/user-attachments/assets/00b195f9-3bea-4680-a338-7d477a40c8ba" />

3. Use the subsection as a prerequisite for other subsection
<img width="1128" height="737" alt="image" src="https://github.com/user-attachments/assets/7ad3c3a5-5961-47e0-a284-69659ea16b3d" />


## How to test

`python -Wd -m pytest -p no:randomly --ds=cms.envs.test ./xmodule/modulestore/tests/test_xml_importer.py::TestSequentialPrerequisitesImport::test_gating_api_calls`

`python -Wd -m pytest -p no:randomly --ds=cms.envs.test ./xmodule/tests/test_sequence.py::SequenceBlockTestCase::test_export_sequence_with_prerequisites`


## Linked Issue: https://github.com/openedx/edx-platform/issues/36995


## Supporting information
1. https://github.com/openedx/edx-platform/issues/36995
2. https://discuss.openedx.org/t/request-for-comments-course-export-import-and-re-run/16508?u=haftamu.kebede

haftamuk committed:
feat: Subsection(sequential) prerequisites persist during course export/import

     *  feat: adds prerequisite information attributes to exported OLX. e.g.
<sequential
required_content="91d0290972c4488db10d7ca3694e13ca"
min_score="100"
min_completion="100"
... >
    <vertical url_name="some_vertical"/>
   ...
</sequential>

    * feat: parse prerequisite information attributes during import and
create gating relationship.

Issue Link: https://github.com/openedx/edx-platform/issues/36995

haftamuk committed:
feat: Subsection(sequential) prerequisites persist during course export/import

     *  feat: adds prerequisite information attributes to exported OLX. e.g.
<sequential
required_content="91d0290972c4488db10d7ca3694e13ca"
min_score="100"
min_completion="100"
... >
    <vertical url_name="some_vertical"/>
   ...
</sequential>

    * feat: parse prerequisite information attributes during import and
create gating relationship.

Issue Link: https://github.com/openedx/edx-platform/issues/36995

tecoholic reviewed (changes requested):
@haftamuk Hi, the changes look good on the export side. I was able to verify that the `<prerequisite>` block is added to the output XML as expected. However, when I tried to reimport it in a new course, I ended up with this:

<img alt="image" src="https://github.com/user-attachments/assets/5cb5defa-18ad-4a1b-89e8-2c4b2c405b7d" />

So, we are clearly missing the "import" part of the solution. Can you kindly add that functionality?

Additionally, I would recommend adding some unit test around the newly created functions. This is a very "core" part of the system and it's best to have unit tests in place.

haftamuk commented:
> So, we are clearly missing the "import" part of the solution. Can you kindly add that functionality?

Kindly see the import functionality is added!

haftamuk commented:
> Additionally, I would recommend adding some unit test around the newly created functions. This is a very "core" part of the system and it's best to have unit tests in place.

@tecoholic I will proceed to ensure existing unit tests pass and introduce new unit tests for the main changes made.
Thank you.

tecoholic commented:
@haftamuk Thanks for updating the PR with the import implementation. It looks like you have used a Generative AI / LLM for the implementation and hence can't be used as is. Kindly consult the [[official policy on use of LLMs](https://openedx.atlassian.net/wiki/spaces/COMM/pages/5022416899/Open+edX+Policy+for+Generative+AI+Tools) for acceptable usage guidelines. In short, you can use it to experiment and create solutions, but the final code should be from the developer.

That doesn't mean we will have to discard this work. Let's use this as a base for the final solution. 

I have an important request - for future work on this PR, please **do not force-push changes**. This will allow us to work ahead incrementally and rollback things if we end up doing something we don't want in the final solution. Kindly see my inline comments in the diff for specific details.

tecoholic commented on cms/djangoapps/contentstore/tasks.py:
```diff
@@ -732,7 +995,7 @@ def read_chunk():
 
         courselike_items = import_func(
-            modulestore(), user.id,
+            modulestore(), user_id,
```
Kindly rever this change. This has nothing to do with the functionality we are working on.

tecoholic commented on xmodule/modulestore/xml.py:
```diff
@@ -65,10 +65,8 @@ def __init__(self, xmlstore, course_id, course_dir,  # lint-amnesty, pylint: dis
-            that xml.
-            """
+        def process_xml(xml, source_info="unknown"):
+            """Takes an xml string, and returns a XBlock created from that xml."""
```
The `source_info` doesn't seem to be used anywhere. This change should be reverted.

tecoholic commented on xmodule/modulestore/xml.py:
```diff
@@ -158,13 +156,16 @@ def fallback_name(orig_name=None):
 
             try:
+                # DEBUG: Log the original XML content
+
```
Nit: unnecessary comments.

tecoholic commented on cms/djangoapps/contentstore/tasks.py:
```diff
@@ -769,10 +1037,10 @@ def read_chunk():
     if is_course:
         sync_discussion_settings(courselike_key, user)
 
-
```
There are a number of places where empty lines are being added and removed. Kindly revert those changes.

tecoholic commented on xmodule/modulestore/xml.py:
```diff
@@ -381,7 +443,9 @@ class to use if none is specified in entry_points
                                   os.path.exists(self.data_dir / d / self.parent_xml)])
+        logging.info(f"MOOVE: XMLModuleStore initializing with source_dirs: {source_dirs}")
         for course_dir in source_dirs:
+            logging.info(f"MOOVE: Loading course from directory: {course_dir}")
```
These 2 logging statements seem to have no relationship with the rest of the changes.

tecoholic commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -524,16 +521,12 @@ def depth_first(subtree):
         depth_first(source_courselike)
-
         for leftover in all_locs:
-            if self.verbose:
-                log.debug('importing block location %s', leftover)
-
```
Why are we removing the logging statments here?

tecoholic commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -548,7 +541,6 @@ def depth_first(subtree):
                 log.exception(
                     f'Course import {dest_id}: failed to import block location {leftover}'
                 )
-                # pylint: disable=raise-missing-from
```
I think we should leave this pylint exception in place.

tecoholic commented on xmodule/seq_block.py:
```diff
@@ -70,6 +70,44 @@ class SequenceFields:  # lint-amnesty, pylint: disable=missing-class-docstring
+        default={},
+        scope=Scope.settings
+    )
+
```
The pre-requisite information is completely managed by the `GatingService` the XBlocks get that information only during the rendering time. So storing them here like this means, we are duplicating this data and would cause multiple problems.

So these fields should be needed at all. This also affects the import/export functionality implemented in this PR. For example, if you export a course -> import it -> then reexport it, you get 2 different XMLs.

<img width="1154" height="447" alt="image" src="https://github.com/user-attachments/assets/1cf517be-d5c0-42e8-b04e-0933e4d08995" />

So, the pre-req relationship should always be established with the help of the GatingService. If you run into issues trying to implement it that way, let's discuss alternates.

tecoholic commented on xmodule/xml_block.py:
```diff
@@ -139,6 +138,65 @@ class XmlMixin:
+            prereq_element.append(min_completion_element)
+        # Insert at the beginning
+        xml_object.insert(0, prereq_element)
+
```
I took some time to understand the functionality of pre-requisites and the `GatingService`. This implementation correctly captures all the necessary information. However, I have a suggestion. The current XML seems verbose.

```
  <prerequisite>
    <required_sequential url_name="e239e393391849c78492e66187af1491"/>
    <min_score>50</min_score>
    <min_completion>80</min_completion>
  </prerequisite>
```

I think we can achive the same functionality with

```
<prerequisite required_content="Usage Key as a string" min_score="value" min_completion="value" />
```

It could result in a simpler export implementation as well. I am a 100% on the "Usage key as a string" in my suggestion. My intention here is to make it generic and not make the assumption that the pre-requisite will always be another "sequential". It could be other things in the future. So, I am thinking if we use the entire usage key, we will capture the necessary information. If we find that's problematic, we can do something like `<prereqisite required_content_key="abc" content_type="sequential" .... />` in the next iteration.

What do you think?

haftamuk replied:
I think you are correct. Kindly have a look at the update!

tecoholic commented on xmodule/xml_block.py:
```diff
@@ -442,6 +500,9 @@ def add_xml_to_node(self, node):
             return
 
+        if self.category == 'sequential':
+            self._add_prerequisite_to_xml(xml_object)  # Uncomment this later
```
```suggestion
            self._add_prerequisite_to_xml(xml_object)
```

tecoholic commented on cms/djangoapps/contentstore/tasks.py:
```diff
@@ -529,11 +532,271 @@ def sync_discussion_settings(course_key, user):
+        return True
+
+    except Exception:
+        return False
```
This looks like a lot of LLM generated code. So, I am not going to be reviewing it deeply. Howerver, I would like to point out a couple of things that relates to the functionality implemented here that will help clean this up.

This `tasks.py` file is just the place where the background task for importing the course is defined. The **actual importing** is done elsewhere. The imports section imports it like

```
from xmodule.modulestore.xml_importer import CourseImportException, import_course_from_xml, import_library_from_xml
```

So, our changes should ideally go into the `import_course_from_xml`. That function uses the [CourseImportManager](https://github.com/openedx/edx-platform/blob/81fc0bd85ce9b5361839900378ee8400fa084e95/xmodule/modulestore/xml_importer.py#L821-L827) to actually perform the import. This is where I would start looking to find the right place to make the changes.

tecoholic replied:
With respect the code itself, the LLM is super-super-super defensive and uses an excessive amount of "try..except" blocks, because it's tuned to produce a "working code". So, I would recomment removing all the `try...except` blocks, just get the bare bones functionality first and then go from there. This would allow us to manually catch those exceptions during development and understand potential issues in the implementation.

For example, there is a ` # Try multiple times with delays (in case of timing issues)` section. Why do we need that? The only reason we need that would be if our processing is time dependent in some way. For example, making a HTTP request that can timeout. In our case, we are simply processing a bunch of XML documents locally.

But LLM gets the `gating_api` and `gating_service` in it's context, and produces code related to HTTP API request and retry logic.

Almost all of the code here is probably useless, it completely relies on the newly introduced properties like `prerequisite` on `SequenceFields`. Some of the things will be useful when implementing the import logic in the XML Importer, like transforming urls to UsageKeys, verifying that the sequentials exist (i.e., they were successfully imported) before establishing the pre-requisite relationship.

tecoholic commented:
@haftamuk I realized, there is lot of things I have noted in comments. So, let's go step by step in addressing them

1. Just do a pass on cleaning up the single line changes like empty lines, comments..etc.,
2. Next, let's focus on the export XML format - [this comment](https://github.com/openedx/edx-platform/pull/37475#discussion_r2459023003)

Ping me for a review after that. Let's move on to the import part of the changes after that.

tecoholic commented on openedx/core/lib/gating/services.py:
```diff
@@ -54,3 +55,57 @@ def is_gate_fulfilled(self, course_key, gating_content_key, user_id):
+        Returns:
+            dict or None: The gating milestone dict or None
+        """
+        return gating_api.get_required_content(course_key, gated_content_key)
```
@haftamuk I took a look into what this is doing and it seems to be the same as the [existing function `required_prereq`](https://github.com/openedx/edx-platform/blob/6deb4f8d0514f88d1c02407ee8554f89da5abd74/openedx/core/lib/gating/services.py#L28-L28).

We can achieve thiis function's functionality with `gating_service.required_prereq(course_key, content_key, 'requires')` in `xml_block.py` and drop this.

tecoholic replied:
On second thought, I think this is fine. Because `required_prereq` simply returns a dict of the milestone, whereas `get_required_content` returns the parsed values of the `content_key`, `min_score` and `min_completion`, which is convenient for the usecase and also prevents duplicating the dict parsing logic.

tecoholic commented on xmodule/xml_block.py:
```diff
@@ -139,6 +140,42 @@ class XmlMixin:
+        if not prereq_info or not prereq_info[0]:
+            return
+        prereq_id, min_score, min_completion = prereq_info
+        prereq_url_name = self._extract_prereq_url_name(prereq_id)
```
@haftamuk The `_extract_prereq_url_name` is a single `if/else` satetement. So maybe we can simply do that check inline here?

I took a look athe possible values of `prereq_id` that the gating service could return. There are 3 possible values - [`None`](https://github.com/openedx/edx-platform/blob/6deb4f8d0514f88d1c02407ee8554f89da5abd74/openedx/core/lib/gating/api.py#L288-L288), [empty string, valid key as string](https://github.com/openedx/edx-platform/blob/6deb4f8d0514f88d1c02407ee8554f89da5abd74/openedx/core/lib/gating/api.py#L422-L427).

So, we can simply do

```
if not prereq_id:
    return
prereq_url_name = UsageKey.from_string(...
```

tecoholic replied:
@haftamuk With this I think we are in a decent shape, when it comes to exporting the content. So, I started looking at the import side of things. It took me to the [process_xml](https://github.com/openedx/edx-platform/blob/6deb4f8d0514f88d1c02407ee8554f89da5abd74/xmodule/modulestore/xml.py#L68-L68) function of the xml ImportSystem. And that seems to consider all the [nodes of an XML document as an XBlock](https://github.com/openedx/edx-platform/blob/master/xmodule/modulestore/xml.py#L163-L163).

That would mean, if we put a `<prerequisite>` tag inside the `<sequential>` that would mean `prerequisite` is an XBlock - which is not the case. Furthermore, [the OLX docs](https://docs.openedx.org/en/latest/educators/olx/organizing-course/course-xml-file.html#sequential-children) also says

> The `sequential` element contains one or more child `vertical` elements.

So, we can conclude that adding a new `<prerequisite>` tag is not the correct way to store this information.

Looking at the [docs on `sequential`'s attributes](https://docs.openedx.org/en/latest/educators/olx/organizing-course/course-xml-file.html#sequential-attributes), it seems like all the other metadata like grading, graceperiod, showanswer ..etc., are stored as attributes on the `<sequential>` tag itself. That would be the correct way to store the pre-requisite information as well.

Can you kindly change the implementation so that the values are set on the `xml_object` itself instead of adding a new element?

haftamuk commented:
> @haftamuk Thanks for updating the PR with the import implementation. It looks like you have used a Generative AI / LLM for the implementation and hence can't be used as is. Kindly consult the [[official policy on use of LLMs](https://openedx.atlassian.net/wiki/spaces/COMM/pages/5022416899/Open+edX+Policy+for+Generative+AI+Tools) for acceptable usage guidelines. In short, you can use it to experiment and create solutions, but the final code should be from the developer.
> 
> That doesn't mean we will have to discard this work. Let's use this as a base for the final solution.
> 
> I have an important request - for future work on this PR, please **do not force-push changes**. This will allow us to work ahead incrementally and rollback things if we end up doing something we don't want in the final solution. Kindly see my inline comments in the diff for specific details.

I am only using AI to experiment and explore the code base. I have cleaned it all now! @tecoholic Kindly have a look at the recent changes and let me know your feedback.

Thank you

haftamuk commented:
> @haftamuk I realized, there is lot of things I have noted in comments. So, let's go step by step in addressing them
> 
> 1. Just do a pass on cleaning up the single line changes like empty lines, comments..etc.,
> 2. Next, let's focus on the export XML format - [this comment](https://github.com/openedx/edx-platform/pull/37475#discussion_r2459023003)
> 
> Ping me for a review after that. Let's move on to the import part of the changes after that.

@tecoholic Thank you very much for the close followup. I think it is now in a better shape as we manage to get close to the critical code components that handle import/export. Kindly have a look at the recent changes and let me know.

I have changed the status of the PR to draft as I thought it would be better to open it once positive feedback is given.

Thank you

tecoholic commented:
@haftamuk I tested the latest changes and the export doesn't work and I know why that is.

* The export is created by the Celery Task, which [uses the modulestore](https://github.com/openedx/edx-platform/blob/12f0a7bbeb4a6d12247e2b50d7f7e785152efa19/cms/djangoapps/contentstore/tasks.py#L419)
* That returns a modulestore instance [created with a specific set of services](https://github.com/openedx/edx-platform/blob/a8a8ae3286db2a76d36c7aa6743e1d5f96729c9f/xmodule/modulestore/django.py#L252)
* The `gating` service is not one of them. So `self.runtime.service('gating')`, [in this line](https://github.com/openedx/edx-platform/pull/37475/files#diff-8de5e71ac655c41107576b84ce0f7b8af02eb7164e80386645b8a1f14ada9cebR159) is always going to be `None`.
* The runtime that's being used during the export is the `SplitModuleStoreRuntime`. It seems to get [the services from the `SplitMongoModuleStore`](https://github.com/openedx/edx-platform/blob/834cb9482d4612dc24ada1bdd53c5f8bdc7eb8fb/xmodule/modulestore/split_mongo/split.py#L675) which is again initialized with the specific set of services mentioned above.

Now the question is, **how can we get the gating service during the export?**

For that, I don't have a good understanding. I can see that  [additional  services are added when the runtime is created](https://github.com/openedx/edx-platform/blob/834cb9482d4612dc24ada1bdd53c5f8bdc7eb8fb/xmodule/modulestore/split_mongo/split.py#L3290). So maybe we can do the same. But let's get the opinion of someone with a bit more experience with this part of the code.

@Agrendalath Hi, you are the first person I can think of who has a good understanding of the modulestore and runtimes. Can you kindly help here? What would you recommend as the right way to get the gating information in the `XmlBlock` during exports?

Agrendalath commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +836,34 @@ def import_library_from_xml(*args, **kwargs):
     return list(manager.run_imports())
 
 
+def _process_sequantial_prerequisites(block):
```
Typo: `sequantial` in this whole file.

Agrendalath commented on xmodule/xml_block.py:
```diff
@@ -139,6 +140,40 @@ class XmlMixin:
+        Add prerequisite information to sequential XML for export.
+        """
+        if self.category != 'sequential':
+            return
```
Why are we modifying the whole `XmlMixin` instead of overriding the `add_xml_to_node` method in the `SequenceBlock`?

Agrendalath commented on xmodule/xml_block.py:
```diff
@@ -139,6 +140,40 @@ class XmlMixin:
+        """
+        if self.category != 'sequential':
+            return
+        gating_service = self.runtime.service(self, 'gating')
```
@haftamuk, do we even need the `GatingService` here? I don't fully understand why we want to proxy calling the `gating_api.get_required_content` function via the `get_required_prereq_metadata` method of a stateless service (instead of calling `gating_api.get_required_content` here directly). We do not need the runtime for this call.

cc: @tecoholic

Agrendalath commented on xmodule/xml_block.py:
```diff
@@ -443,6 +478,9 @@ def add_xml_to_node(self, node):
             return
 
+        if self.category == 'sequential':
+            self._add_prerequisite_to_xml(xml_object)
```
Aside from moving this to the `SequenceBlock`, we should also check [whether the gating is enabled](https://github.com/openedx/edx-platform/blob/7b953a5310ccd51d24207b70a17844ce3c83e664/openedx/core/lib/gating/api.py#L81-L100).

Agrendalath reviewed (commented):
@tecoholic, I skimmed through the code and left some notes. I did not do the full review, though, so I may not have the full context.

cc: @haftamuk

haftamuk commented:
> @tecoholic, I skimmed through the code and left some notes. I did not do the full review, though, so I may not have the full context.
> 
> cc: @haftamuk

@Agrendalath  we have gone through your comments and did responded. The export is now working reliably (Because with the GatingService it was not reliable sometimes working sometimes not!)

Thank you for the quick feedback.

haftamuk commented:
> @haftamuk I tested the latest changes and the export doesn't work and I know why that is.
> 
> * The export is created by the Celery Task, which [uses the modulestore](https://github.com/openedx/edx-platform/blob/12f0a7bbeb4a6d12247e2b50d7f7e785152efa19/cms/djangoapps/contentstore/tasks.py#L419)
> * That returns a modulestore instance [created with a specific set of services](https://github.com/openedx/edx-platform/blob/a8a8ae3286db2a76d36c7aa6743e1d5f96729c9f/xmodule/modulestore/django.py#L252)
> * The `gating` service is not one of them. So `self.runtime.service('gating')`, [in this line](https://github.com/openedx/edx-platform/pull/37475/files#diff-8de5e71ac655c41107576b84ce0f7b8af02eb7164e80386645b8a1f14ada9cebR159) is always going to be `None`.
> * The runtime that's being used during the export is the `SplitModuleStoreRuntime`. It seems to get [the services from the `SplitMongoModuleStore`](https://github.com/openedx/edx-platform/blob/834cb9482d4612dc24ada1bdd53c5f8bdc7eb8fb/xmodule/modulestore/split_mongo/split.py#L675) which is again initialized with the specific set of services mentioned above.
> 
> Now the question is, **how can we get the gating service during the export?**
> 
> For that, I don't have a good understanding. I can see that [additional services are added when the runtime is created](https://github.com/openedx/edx-platform/blob/834cb9482d4612dc24ada1bdd53c5f8bdc7eb8fb/xmodule/modulestore/split_mongo/split.py#L3290). So maybe we can do the same. But let's get the opinion of someone with a bit more experience with this part of the code.
> 
> @Agrendalath Hi, you are the first person I can think of who has a good understanding of the modulestore and runtimes. Can you kindly help here? What would you recommend as the right way to get the gating information in the `XmlBlock` during exports?

@tecoholic Thanks for the detailed explanation and pulling @Agrendalath 

Now, as we are using the GatingService api methods directly, it seems reliably working.

@tecoholic and @Agrendalath kindly if you can go through the modifications (its now just two files!) and put your thoughts before progressing to the next step.

Thank you very much.

tecoholic commented:
@Agrendalath Thank you for taking the time to review this and thanks @haftamuk for quickly addressing the comments. I will give it another test run. If you find things working realiabily, I think we can go ahead with some unit tests as well.

@Agrendalath I would like to learn more about [your comment here](https://github.com/openedx/edx-platform/pull/37475#discussion_r2519196111). How do we determine the usage of runtime vs directly using the API? You mentioned, we can directly use the API, but the `SeqBlock` already has a `XBlock.wants('gating')` decorator on it, so wouldn't we be able to use the service now? Why would we be importing the gating_api directly?

haftamuk commented:
> @tecoholic, I skimmed through the code and left some notes. I did not do the full review, though, so I may not have the full context.
> 
> cc: @haftamuk

Also replying based on the latest changes that definition_to_xml is overridden  in SequenceBlock which is called in add_xml_to_node. Following a discussion today with @Cup0fCoffee we thought this is a better done at definition_to_xml instead of overriding add_xml_to_node which resulted in more complicated code.

Also unit tests are added to verify prerequisite data is preserved during import and export.
@tecoholic and @Agrendalath for your input on the latest changes!

Thank you very much.

Agrendalath commented:
@tecoholic,
> @Agrendalath I would like to learn more about https://github.com/openedx/edx-platform/pull/37475#discussion_r2519196111. How do we determine the usage of runtime vs directly using the API? You mentioned, we can directly use the API, but the SeqBlock already has a XBlock.wants('gating') decorator on it, so wouldn't we be able to use the service now? Why would we be importing the gating_api directly?

There are a few different aspects to consider:
1. For the XML export, we use a simplified runtime that's not bound to a user (i.e., not something like the [prepare_runtime_for_user](https://github.com/openedx/edx-platform/blob/7b953a5310ccd51d24207b70a17844ce3c83e664/lms/djangoapps/courseware/block_render.py#L487) function that [would initialize most services](https://github.com/openedx/edx-platform/blob/7b953a5310ccd51d24207b70a17844ce3c83e664/lms/djangoapps/courseware/block_render.py#L589-L640) in the student view scenario). It means that the services are limited to the [basic ones](https://github.com/openedx/edx-platform/blob/7b953a5310ccd51d24207b70a17844ce3c83e664/xmodule/modulestore/split_mongo/split.py#L675-L685) or they need to be added [in some hacky way](https://github.com/openedx/edx-platform/blob/7b953a5310ccd51d24207b70a17844ce3c83e664/xmodule/modulestore/split_mongo/runtime.py#L80).
2. That said, we basically have two types of services:
2.1. The ones that store some runtime state, like [the user data](https://github.com/openedx/edx-platform/blob/7b953a5310ccd51d24207b70a17844ce3c83e664/lms/djangoapps/courseware/block_render.py#L592-L602) or [the current course ID](https://github.com/openedx/edx-platform/blob/7b953a5310ccd51d24207b70a17844ce3c83e664/lms/djangoapps/courseware/block_render.py#L637).
2.2. The ones that do not get any data on init, which is the case for [GatingService](https://github.com/openedx/edx-platform/blob/7b953a5310ccd51d24207b70a17844ce3c83e664/openedx/core/lib/gating/services.py#L8). Their main purpose is to provide a unified interface that can be used by XBlocks. Since we do not need to reuse this content gating logic for other XBlocks, it's okay to call the `gating_api` directly in this case.

tecoholic commented on xmodule/seq_block.py:
```diff
@@ -978,8 +978,39 @@ def definition_to_xml(self, resource_fs):
+        if not prereq_info or not prereq_info[0]:
+            return
+        prereq_id, min_score, min_completion = prereq_info
+        prereq_url_name = self._extract_prereq_url_name(prereq_id)
```
The `_extract_prereq_url_name` isn't doing anything complex and it's also used only once here. So, I think we can get rid of it and add the logic directly here.

tecoholic commented on xmodule/seq_block.py:
```diff
@@ -978,8 +978,39 @@ def definition_to_xml(self, resource_fs):
+            self.location.course_key,
+            self.location
+        )
+        if not prereq_info or not prereq_info[0]:
```
Let's add `log.debug` statements for each of these `if condition: return` scenarios. It would be helpful for anyone tryint to debug later.

tecoholic commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +836,39 @@ def import_library_from_xml(*args, **kwargs):
+def _process_sequential_prerequisites(block):
+    """
+    extracts sequential/subsection prerequisite information, enables make required
+    for a required sequential and creates a gating for the target sequential.
```
```suggestion
    Extracts prerequisite information for sequential blocks, and sets up
    the gating relationship.
```

tecoholic commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -918,6 +953,10 @@ def _convert_ref_fields_to_new_namespace(reference):
     )
 
+    # Utilize sequential pre-requisites information to create gating relationship
+    if block.location.block_type == "sequential":
+        _process_sequential_prerequisites(block)
```
Add a gating enabled check mentioned in [this comment](https://github.com/openedx/edx-platform/pull/37475#discussion_r2519210398)

Agrendalath replied:
On the second thought, it might be okay to run it every time, as the prerequisites may exist for subsections even when they are not enabled in the course.

tecoholic commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +836,39 @@ def import_library_from_xml(*args, **kwargs):
+        return
+    sequential_prerequisite_attrs = ['required_content', 'min_score', 'min_completion']
+    if not all(attr in block.xml_attributes for attr in sequential_prerequisite_attrs):
+        return
```
Instead of checking if these attributes exist, can we first call `get()` method with a default value and then perform the validation on the values? So, instead of 

```py
if not attr in block.xml_attributes:
    return
value = block.xml_attributes.get(attr)
gating_api.function(value) # <-- here we assume the value is correct
```

how about we do something like this

```py
try:
    required_content = block.xml_attributes.get(attr, "")
    min_score = int(block.xml_attributes.get("min_score", None))
except ValueError:
    log_and_return

if not value1 or not value2:
    log_and_return

gating_api.function(required_conent, min_score,...)  # <-- validated values that we know are valid
```

tecoholic commented on cms/djangoapps/contentstore/views/tests/test_import_export.py:
```diff
@@ -1055,6 +1176,59 @@ def test_export_status_handler_filesystem(
+        )
+
+        # Export the course
+        tar_ball = self.test_export_async()
```
This testcase is very extensive, as it tests the full import. This is more of an integration test than a unit test. Let us simplify this test case and add it to the `SequenceBlockTestCase` in `xmodule/tests/test_sequence.py`. Ideally just testing the `_add_prerequisite_to_xml` function. The test case already has the necessary `setUp` functions to create the mock course blocks..etc., so, it should be a relatively small test case compared to this.

tecoholic commented on cms/djangoapps/contentstore/views/tests/test_import_export.py:
```diff
@@ -732,6 +732,127 @@ def test_import_status_response_is_not_cached(self, fmt):
+            )
+
+            # Import the course as usual
+            import_course_from_xml(
```
Similar to the export tests below, this is also more of an integration test than a unit test. So, let's simplify this and move it to `xmodule/moudulestore/tests/test_xml_importer.py`. I don't see a test case that set's up everything like the one for the export function. So, we might have to create a new test case there, mock the sequential block that we are passing to the `_process_sequential_prerequisites` and verify that it behaves as expected.

tecoholic reviewed (changes requested):
@haftamuk Functionality wise, I think we are good. I have left some comments about code organization. Specifically the tests. Can you kindly address them?

tecoholic commented:
@Agrendalath Thank you for explaining things in detail.

haftamuk commented:
Hello @Agrendalath ,
If  you have time, if you can have a look at the PR if it is ready to merge!

Thank you very much.

Agrendalath commented:
@tecoholic , would you be able to review this? I can review it as the CC in the next sprint, but I still have a few things to address in the current one.

cc: @haftamuk

tecoholic commented:
@Agrendalath Absolutely. I will review it this sprint and you can do the CC in next.

tecoholic commented on xmodule/modulestore/tests/test_xml_importer.py:
```diff
@@ -257,3 +258,48 @@ def test_import_static_file(self):
+        and calls gating_api functions add_prerequisite and set_required_content with valid parameters
+        """
+        self.mock_sequential.xml_attributes = {
+            'required_content': 'gated_sequential',
```
This test is confusing because the `self.sequential_location` is currently set to `gated_sequential`, and the `required_content` here is set to `gated_sequential` as well. This should be changed to something like `required_sequential`

tecoholic commented on xmodule/modulestore/tests/test_xml_importer.py:
```diff
@@ -257,3 +258,48 @@ def test_import_static_file(self):
+
+            mock_gating_api.add_prerequisite.assert_called_once_with(
+                self.course_key,
+                self.course_key.make_usage_key('sequential', 'gated_sequential')
```
.. and then this would become
```suggestion
                self.course_key.make_usage_key('sequential', 'required_sequential')
```

tecoholic commented on xmodule/modulestore/tests/test_xml_importer.py:
```diff
@@ -257,3 +258,48 @@ def test_import_static_file(self):
+            mock_gating_api.set_required_content.assert_called_once_with(
+                self.course_key,
+                self.sequential_location,
+                self.course_key.make_usage_key('sequential', 'gated_sequential'),
```
.. and this becomes
```suggestion
                self.course_key.make_usage_key('sequential', 'required_sequential'),
```

tecoholic commented on xmodule/modulestore/tests/test_xml_importer.py:
```diff
@@ -257,3 +258,48 @@ def test_import_static_file(self):
+        self.mock_sequential = mock.Mock()
+        self.mock_sequential.location = self.sequential_location
+
+    def test_gating_api_calls(self):
```
This tests only the happy path where the the mock_sequential has all the necessary values. More test cases convering the other code paths like the `if` conditional, the `try...except` block where the values are either missing or not in the expected format (int, string..etc) should be added to this test case.

tecoholic commented on xmodule/tests/test_sequence.py:
```diff
@@ -572,3 +573,30 @@ def test_prereqs_met_content_rendered_normally(self):
+        # content prepared here will be returned. Therefore, verification of
+        # contents in the exported olx/xml to those values will be verified.
+        mock_get_required_content.return_value = (
+            "block-v1:TestX+TestCourse+1+type@sequential+block@gated_sequential",
```
```suggestion
            "block-v1:TestX+TestCourse+1+type@sequential+block@required_sequential",
```

tecoholic commented on xmodule/tests/test_sequence.py:
```diff
@@ -572,3 +573,30 @@ def test_prereqs_met_content_rendered_normally(self):
+            self.sequence_3_1.location
+        )
+
+        assert xml_object.get('required_content') == 'gated_sequential'
```
```suggestion
        assert xml_object.get('required_content') == 'required_sequential'
```

tecoholic commented on xmodule/tests/test_sequence.py:
```diff
@@ -572,3 +573,30 @@ def test_prereqs_met_content_rendered_normally(self):
         assert metadata["prev_url"] == "prev_url"
+
+    @patch("openedx.core.lib.gating.api.get_required_content")
+    def test_export_sequence_with_prerequisites(self, mock_get_required_content):
```
Similar to the importer testcase, this only checks the happy path of the code. Let's add test cases where the 'if' conditions in `add_prerequisite_to_xml` produce other resutls as well.

tecoholic reviewed (changes requested):
@haftamuk This is looking great. I have a few more asks on the tests department. Can you kindly address those?

tecoholic commented on xmodule/tests/test_sequence.py:
```diff
@@ -572,3 +573,123 @@ def test_prereqs_met_content_rendered_normally(self):
+        # Create a mock XML element
+        xml_object = etree.Element('sequential')
+
+        # When add_prerequisite_to_xml method calls get_required_content mocked
```
Nit: duplicate comments

tecoholic commented on xmodule/tests/test_sequence.py:
```diff
@@ -572,3 +573,123 @@ def test_prereqs_met_content_rendered_normally(self):
+        )
+        assert xml_object.get('required_content') is None
+        assert xml_object.get('min_score') is None
+        assert xml_object.get('min_completion') is None
```
I think it would be better to use `assert 'required_content' not in xml_object.attrib` to verify that these attibutes are actually not added to the xml object. Current this test doesn't distinguish between a non-existant attribute and a attribute with 'None' value. See 'b' and 'c' in this example

```
>>> from xml.etree.ElementTree import Element
>>> a = Element('a')
>>> a.set('b', None)
>>> a
<Element 'a' at 0x7819e607d260>
>>> a.attrib
{'b': None}
>>> a.get('b')
>>> a.get('c')
>>>
```

Ref: https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.attrib

tecoholic reviewed (approved):
@haftamuk :+1: 

- [x] I tested this: I verified that the export and import of courses preserves pre-requisites in sequentials.
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation - Handled in a separate PR.

haftamuk commented:
Hello @Agrendalath,
I am just trying to pull this to your visibility for a CC review.

Thank you

tecoholic commented:
@Agrendalath Hi, a gentle reminder on taking a look at this PR.

Agrendalath commented:
Sorry for the delay, @haftamuk, @tecoholic. I will review this later this week.

haftamuk commented:
Hello @Agrendalath ,
i did rebased the PR branch on top of the recent changes. Also bringing this to top as a reminder.

Thank you.

Agrendalath commented on xmodule/seq_block.py:
```diff
@@ -978,8 +978,43 @@ def definition_to_xml(self, resource_fs):
+        prereq_id, min_score, min_completion = prereq_info
+        if (
+            prereq_id is None
+            or not isinstance(prereq_id, str)
```
Nit: the second condition already covers the first one.

Agrendalath commented on xmodule/seq_block.py:
```diff
@@ -978,8 +978,43 @@ def definition_to_xml(self, resource_fs):
+        self.add_prerequisite_to_xml(xml_object)
         return xml_object
 
+    def add_prerequisite_to_xml(self, xml_object):
```
The exports do not contain the information whether a subsection is available as a prerequisite: https://github.com/openedx/edx-platform/blob/d6633243a8397bb56eb21e38cc43efada4f05931/cms/templates/js/access-editor.underscore#L43-L55

Agrendalath replied:
@haftamuk, a reminder about [this one](https://github.com/openedx/edx-platform/pull/37475#discussion_r2607401937). It should be reflected here and in the importer.

Agrendalath commented on xmodule/seq_block.py:
```diff
@@ -978,8 +978,43 @@ def definition_to_xml(self, resource_fs):
+            return
+        if (
+            min_score is None
+            or not isinstance(min_score, int)
```
`min_score` and `min_completion` can be set to an empty or invalid value in Studio, in which case the API will return an empty string. The exporting function should not be responsible for validating them.

Agrendalath commented on xmodule/seq_block.py:
```diff
@@ -978,8 +978,43 @@ def definition_to_xml(self, resource_fs):
+            self.location.course_key,
+            self.location
+        )
+        prereq_id, min_score, min_completion = prereq_info
```
```suggestion
        prereq_usage_key_str, min_score, min_completion = prereq_info
```

The `id` is unclear here.

Agrendalath commented on xmodule/seq_block.py:
```diff
@@ -978,8 +978,43 @@ def definition_to_xml(self, resource_fs):
+            prereq_id is None
+            or not isinstance(prereq_id, str)
+        ):
+            log.debug('Unable to extract Usage key: %s', prereq_info)
```
Do we want to log any of these? These logs are not particularly helpful for debugging, and they would be triggered for all subsections that lack prerequisites.

Agrendalath commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +835,51 @@ def import_library_from_xml(*args, **kwargs):
+        block: The sequential block
+    """
+    if not hasattr(block, "xml_attributes") or not block.xml_attributes:
+        log.debug('Block has no xml_attributes property: %s', block)
```
Similar to [this comment](https://github.com/openedx/edx-platform/pull/37475/changes#r2607440631) - it would be triggered for all subsections, so it would mostly clutter the logs.

Agrendalath commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +835,51 @@ def import_library_from_xml(*args, **kwargs):
+
+        min_score = block.xml_attributes.get('min_score')
+        min_completion = block.xml_attributes.get('min_completion')
+        if not min_score or not min_completion:
```
How do prerequisites behave when the value is not an integer? Do they still work?

Agrendalath commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +835,51 @@ def import_library_from_xml(*args, **kwargs):
+
+    prerequisite_usage_key = course_key.make_usage_key('sequential', required_content)
+
+    gating_api.add_prerequisite(course_key, prerequisite_usage_key)
```
Let's consider the opposite scenario: the existing sequential in the course has prerequisites, but its equivalent in the course export does not. Should we delete the prerequisites in such a case?

Agrendalath commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +835,51 @@ def import_library_from_xml(*args, **kwargs):
+        prerequisite_usage_key,
+        min_score,
+        min_completion
+    )
```
This will raise an exception if the subsection does not exist. It should be okay to mark the import as failed in this scenario (instead of silently ignoring the errors), but we should ensure that the error message is clear to the course author.

Agrendalath reviewed (changes requested):
Hi @haftamuk, sorry for the delay. I reviewed non-test code and left some comments.

haftamuk committed:
feat: Applyig code review recomendations

This commit contans fixes as per review recomendations. This commit
changes will be merged later after review approval

haftamuk committed:
feat: Applyig code review recomendations

This commit contans fixes as per review recomendations. This commit
changes will be merged later after review approval

Agrendalath commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +836,49 @@ def import_library_from_xml(*args, **kwargs):
+        if not min_score or not min_completion:
+            return
+
+        min_score = int(min_score)
+        min_completion = int(min_completion)
```
This (+catching the exceptions) is redundant. `set_required_content` already validates casting these variables to integers with `_validate_min_score` and raises `GatingValidationError`.

Agrendalath commented on xmodule/seq_block.py:
```diff
@@ -978,8 +978,27 @@ def definition_to_xml(self, resource_fs):
+            self.location.course_key,
+            self.location
+        )
+        prereq_usage_key_str, min_score, min_completion = prereq_info
+        if not isinstance(prereq_usage_key_str, str):
+            return
+        prereq_usage_key = UsageKey.from_string(prereq_usage_key_str)
```
@haftamuk, I did not test this one, but wouldn't it be sufficient to add `prereq_usage_key_str` to XML (without building the `UsageKey`?

Agrendalath commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +836,49 @@ def import_library_from_xml(*args, **kwargs):
+            min_completion
+        )
+    except (GatingValidationError) as e:
+        logging.debug('Gating validation error : %s', {e})
```
@haftamuk, a reminder about [this comment](https://github.com/openedx/edx-platform/pull/37475#discussion_r2611809610). Importing the course would still fail silently.

Agrendalath commented on xmodule/modulestore/xml_importer.py:
```diff
@@ -834,6 +836,49 @@ def import_library_from_xml(*args, **kwargs):
+
+    prerequisite_usage_key = course_key.make_usage_key('sequential', required_content)
+    try:
+        gating_api.add_prerequisite(course_key, prerequisite_usage_key)
```
@haftamuk, a reminder about [this comment](https://github.com/openedx/edx-platform/pull/37475#discussion_r2611790956).

Agrendalath reviewed (changes requested):
@tecoholic, I did another pass. The redundant checks are gone, but the crucial edge cases from my previous review remain unaddressed.

cc: @haftamuk

haftamuk committed:
feat: add XML import/export validation

The import implementation collects the gating_information initially and persists at the end:
- This ensures all blocks in the course are imported before dealing with gating_information
- This prevents import order failures (e.g., when the prerequisite block
  appears after the block that references it).
- Missing prerequisite blocks are logged as errors and skipped, rather
  than failing the entire import.

This implementation avoids ordering dependencies(during imprt order of course components is
unpredictable/not sequantial). In this implementation, invalid relationships are
gracefully skipped with an error log instead of crashing the import.
